app/src/main/java/io/customer/example/MainActivity.kt

```diff
@@ -23,24 +24,70 @@ class MainActivity : AppCompatActivity() {
 
         // 2nd way
         makeAsynchronousRequest()
+
+        // log events
+//        makeEventsRequests()
+    }
+
+    private val outputCallback = Action.Callback<Unit> { result ->
+        when (result) {
+            is ErrorResult -> Log.v("ErrorResult", result.error.getDisplayMessage())
+            is Success -> Log.v("Success", "Success")
+        }
+    }
+
+    data class Fol(val a: String, val c: Int)
+
+    private fun makeEventsRequests() {
+        CustomerIo.instance().track(
+            name = "string event",
+            attributes = mapOf(
+                "value" to "string test",
+                "target" to 1
+            )
+        ).enqueue(outputCallback)
+        CustomerIo.instance().track(
+            name = "int event",
+            attributes = mapOf("value" to 1)
+        ).enqueue(outputCallback)
+        CustomerIo.instance().track(
+            name = "long event",
+            attributes = mapOf("value" to 1L)
+        ).enqueue(outputCallback)
+        CustomerIo.instance().track(
+            name = "array event",
+            attributes = mapOf("value" to listOf("1", "2"))
+        ).enqueue(outputCallback)
+        CustomerIo.instance().track(
+            name = "date event",
+            attributes = mapOf("value" to Date())
+        ).enqueue(outputCallback)
+        CustomerIo.instance().track(
+            name = "timestamp event",
+            attributes = mapOf("value" to Date().time)
+        ).enqueue(outputCallback)
+        CustomerIo.instance().track(
+            name = "custom class event",
+            attributes = mapOf("value" to Fol(a = "aa", c = 1))
+        )
     }
 
     private fun makeAsynchronousRequest() {
         CustomerIo.instance()
             .identify(
-                "identifier",
-                attributes = mapOf("email" to IdentityAttributeValue.StringAttribute("sample@email.com"))
-            ).enqueue {
-                when (it) {
-                    is ErrorResult -> Log.v("ErrorResult", it.error.cause.toString())
-                    is Success -> Log.v("Success", "Success")
-                }
-            }
+                identifier = "identifier",
+                attributes = mapOf("email" to "testemail@email.com")
+            ).enqueue(outputCallback)
     }
 
     private fun makeSynchronousRequest() {
         CoroutineScope(Dispatchers.IO).launch {
-            when (val result = CustomerIo.instance().identify("testcoo@email.com").execute()) {
+            when (
+                val result =
+                    CustomerIo.instance()
+                        .identify("sample@email.com@email.com", mapOf("speed" to "slow"))
+                        .execute()
```

Comment:
I am guessing the idea behind using `Action<>.execute()` is for supporting coroutines?

Also, if I am someone using the SDK and I use `execute()`, how do I perform error handling since it is synchronous? 

Lastly, is there a scenario when a customer would ever want to cancel an `Action`? I am wondering why that functionality exists?

Reply:
`Action` provides you three ways to interact 
- `Execute being synchronous and it results back with a `Result<T>` object which is a sealed class, containing either Sucess or Error object.
- `enqueue(callback: Callback<T>)`, Where `Callback`has `fun onResult(result: Result<T>)` method.

So the client is provided with a `Result` object in all cases, except you when you want to ignore `callback`

- For `cancel` probably doesn't make much sense for `tracking` messages but for `transaction` requests that our system supports it's going to be very handy. Plus, in modern tooling you have ability to cancel `coroutine` `disposable` from RX world, and the cancel feature would compliment if anyone wants to clean up resources.

Reply:
Thanks for helping me better understand the `Action` type. It's trying to offer a flexible way for customers to interact with the SDK. 

Since it's already built, I say we can leave it. Otherwise I would suggest keeping `Action` simple with the intention of making *all* SDK functions synchronous in the future like the iOS SDK is moving towards at the moment.